File: 063_Unique_Paths_II/063_2.cpp
#include <iostream>
#include <string>
#include <vector>
#include <climits>
#include <algorithm>
#include <cstring>
using namespace std;

class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        if (obstacleGrid.empty() || obstacleGrid[0][0] == 1)return 0;
        size_t m = obstacleGrid.size(), n = obstacleGrid[0].size();
        obstacleGrid[0][0] = 1;
        for (int i = 1; i < m; ++i) obstacleGrid[i][0] =
                (obstacleGrid[i][0] == 0 && obstacleGrid[i - 1][0] == 1) ? 1 : 0;
        for (int j = 1; j < n; ++j) obstacleGrid[0][j] =
                (obstacleGrid[0][j] == 0 && obstacleGrid[0][j - 1] == 1) ? 1 : 0;
        for (int i = 1; i < m; ++i) {
            for (int j = 1; j < n; ++j) {
                if (obstacleGrid[i][j] == 0) {
                    obstacleGrid[i][j] = obstacleGrid[i - 1][j] + obstacleGrid[i][j - 1];
                } else {
                    obstacleGrid[i][j] = 0;
                }
            }
        }
        return obstacleGrid.back().back();
    }
};


int main() {
    vector<vector<int>> aa(3, vector<int>(3, 0));
    aa[1][1] = 1;
    Solution s;
    cout << s.uniquePathsWithObstacles(aa) << endl;
}

File: 063_Unique_Paths_II/063_2.py
from typing import List

class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        m = len(obstacleGrid)
        n = len(obstacleGrid[0])
        dp = [1] + [0]*n
        for i in range(0,m):
            for j in range(0,n):
                dp[j] = 0 if obstacleGrid[i][j] else dp[j]+dp[j-1]
        return dp[-2]


if __name__ == "__main__":
    aa = [[0 for _ in range(3)] for _ in range(3)]
    aa[1][1] = 1
    s = Solution()
    print(s.uniquePathsWithObstacles(aa))